### PR TITLE
feat: multiple UX improvements to labels

### PR DESCRIPTION
**Pull request by scmmishra**
This PR has the following changes

1. Sort all the labels alphabetically by default
2. Allow fuzzy search in labels
3. Allow creating new labels from the dropdown only for admins
4. Global keyboard shortcut `L` for labels
5. New `kbd` component
6. Add `isActiveElementTypeable` to check if the current active element accepts an input
7. Add types for `jest`

https://www.loom.com/share/d4e141c2d61c47b38dea1b7588a7f365?sid=94481d0e-b6df-4555-8517-45897830765e

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my code
- [ ] I have commented on my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: app/javascript/shared/components/ui/label/LabelDropdown.vue [hunk 1]
```diff
@@ -49,24 +85,59 @@ export default {
...
+      return (
+        this.allowCreation &&
+        this.filteredActiveLabels.length < 3 &&
+        this.search !== ''
+      );
+    },
+
+    parsedSearch() {
+      const labelRegex = /[^a-zA-Z0-9_-]/g;
+      const spaceRegex = /\s+/g;
```
Comment (iamsivin) [inline app/javascript/shared/components/ui/label/LabelDropdown.vue:133]: Can we move it to a helper file? We can reuse it.

File: app/javascript/shared/components/ui/label/LabelDropdown.vue [hunk 2]
```diff
@@ -97,11 +168,35 @@ export default {
...
 <style lang="scss" scoped>
+.dropdown-title-container {
+  display: flex;
+  justify-content: space-between;
+  align-items: center;
+  margin-bottom: var(--space-smaller);
+
+  .text-block-title {
+    flex-grow: 1;
+    margin: 0px;
```
Comment (iamsivin) [inline app/javascript/shared/components/ui/label/LabelDropdown.vue:192]:
```suggestion
    margin: 0;
```

**Review by iamsivin — COMMENTED** [review]
@scmmishra Nice work. I added some comments. PTAL.

File: app/javascript/shared/helpers/KeyboardHelpers.js [hunk 3]
```diff
@@ -127,3 +127,14 @@ export const buildHotKeys = e => {
...
+export const isActiveElementTypeable = e => {
+  const activeElement = e.target ?? document.activeElement;
+
+  return !!(
+    activeElement?.tagName === 'INPUT' ||
+    activeElement?.tagName === 'NINJA-KEYS' ||
+    activeElement?.tagName === 'TEXTAREA' ||
+    activeElement?.contentEditable === 'true'
+  );
+};
```
Comment (iamsivin) [inline app/javascript/shared/helpers/KeyboardHelpers.js:140]: @scmmishra, Could you please take a look at the `onKeyDownHandler(e)` method in `eventListenerMixins`? Currently, it only handles 'INPUT' and 'TEXTAREA', but we can add the rest there and we can remove this function from here.

Comment (scmmishra) [reply app/javascript/shared/helpers/KeyboardHelpers.js:140]: Sounds good! Will do

File: app/javascript/dashboard/routes/dashboard/conversation/labels/LabelBox.vue [hunk 4]
```diff
@@ -77,13 +83,29 @@ export default {
...
   methods: {
     toggleLabels() {
       this.showSearchDropdownLabel = !this.showSearchDropdownLabel;
     },
     closeDropdownLabel() {
       this.showSearchDropdownLabel = false;
     },
+    handleKeyEvents(e) {
+      const keyPattern = buildHotKeys(e);
+      if (keyPattern === 'l' && !isActiveElementTypeable(e)) {
```
[the remark this answers is not in the document]
Comment (scmmishra) [reply app/javascript/dashboard/routes/dashboard/conversation/labels/LabelBox.vue:98]: Is there a need to do that? The buildHotKeys already handles ensuring the any modifier keys are not present. Apart from that the check is quite straightforward and has no special/edge cases.

[the remark this answers is not in the document]
Comment (iamsivin) [reply app/javascript/dashboard/routes/dashboard/conversation/labels/LabelBox.vue:98]: Fine

**Comment by scmmishra** [comment]
@iamsivin @nithindavid this PR is ready for review. PTAL and let me know your thoughts

**Review by iamsivin — COMMENTED** [review]
@scmmishra Can we hide the kbd `l` label from the contact page sidebar? Or can we implement it there? It's not working there.

image

**Review by nithindavid — COMMENTED** [review]
Amazing PR 🔥 Code looks good. I'll test the cases mentioned and will post a loom video towards evening.

File: yarn.lock [hunk 5]
```diff
@@ -15133,6 +15271,11 @@ react-is@^17.0.1:
   resolved "https://registry.yarnpkg.com/react-is/-/react-is-17.0.2.tgz#e691d4a8e9c789365655539ab372762b0efb54f0"
   integrity sha512-w2GsyukL62IJnlaff/nRegPQR94C/XXamvMWmSHRJ4y7Ts/4ocGRmTHvOs8PSE6pB3dWOrD/nueuU5sduBsQ4w==
 
+react-is@^18.0.0:
```
Comment (nithindavid) [inline yarn.lock:15274]: Can we avoid this package, or is it intentional?

**Comment by scmmishra** [comment]
> @scmmishra Can we hide the kbd `l` label from the contact page sidebar? Or can we implement it there? It's not working there. I tested it locally and working as expected. 🔥
> 
Update the PR, thanks for catching this, @iamsivin

File: yarn.lock (hunk 5 above)
Comment (scmmishra) [reply yarn.lock:15274]: Seems like a peer-dependency from jest, we already have a couple of entries for this

https://github.com/chatwoot/chatwoot/blob/533f9f7fe4999cd460a59711003c4a42122f2b54/yarn.lock#L15126-L15134

It's a dev dependency, should not show up in our builds

**Review by iamsivin — COMMENTED** [review]
@scmmishra Could you handle this case as well? In case the label list is empty, we should display a message saying "No labels found" along with the label add button. This should only appear when a search is being performed.


https://github.com/chatwoot/chatwoot/assets/64252451/4ccf6f87-aaf5-48b4-badd-015c7c42cec0

**Comment by scmmishra** [comment]
> @scmmishra Could you handle this case as well? In case the label list is empty, we should display a message saying "No labels found" along with the label add button. This should only appear when a search is being performed.

great idea, will do

**Comment by scmmishra** [comment]
@iamsivin done

CleanShot 2023-06-23 at 15 06 45@2x

**Review by iamsivin — APPROVED** [review]
I have tested it locally and it is working as expected. The code looks good to me.

**Review by nithindavid — APPROVED** [review]
https://www.loom.com/share/8ce5103f0f2743c0870201298d1af7ef

**Pull request merged**